server/src/main/java/io/spine/server/storage/system/SystemAwareStorageFactory.java

```diff
@@ -0,0 +1,125 @@
+import io.spine.server.event.store.EmptyEventStore;
+import io.spine.server.projection.Projection;
+import io.spine.server.projection.ProjectionStorage;
+import io.spine.server.storage.RecordStorage;
+import io.spine.server.storage.StorageFactory;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * An implementation of {@link StorageFactory} which may produce an empty {@link EventStore} for
+ * system context.
+ *
+ * <p>Delegates storage creation to another factory.
+ *
+ * <p>When asked for a {@link EventStore}, creates an {@link EmptyEventStore} if the given context
+ * does not store events. Otherwise, delegates to the associated delegate factory.
+ */
+@Internal
+public final class SystemAwareStorageFactory implements StorageFactory {
+
+    private final StorageFactory delegate;
+
+    private SystemAwareStorageFactory(StorageFactory delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Wraps the given factory into a {@code SystemAwareStorageFactory}.
+     *
+     * @param factory
+     *         the {@link StorageFactory} to use as the delegate for storage construction
+     * @return a new {@code SystemAwareStorageFactory} of the given {@code factory} if it is
+     *         an instance of {@code SystemAwareStorageFactory}
+     */
+    public static SystemAwareStorageFactory wrap(StorageFactory factory) {
+        checkNotNull(factory);
+        return factory instanceof SystemAwareStorageFactory
+               ? (SystemAwareStorageFactory) factory
+               : new SystemAwareStorageFactory(factory);
+    }
+
+    /**
+     * Obtains the wrapped factory.
+     */
+    @VisibleForTesting
+    public StorageFactory delegate() {
+        return delegate;
+    }
+
+    @Override
+    public <I> AggregateStorage<I>
+    createAggregateStorage(ContextSpec context,
+                           Class<? extends Aggregate<I, ?, ?>> aggregateClass) {
+        return delegate.createAggregateStorage(context, aggregateClass);
+    }
+
+    @Override
+    public <I> RecordStorage<I>
+    createRecordStorage(ContextSpec context, Class<? extends Entity<I, ?>> entityClass) {
+        return delegate.createRecordStorage(context, entityClass);
+    }
+
+    @Override
+    public <I> ProjectionStorage<I>
+    createProjectionStorage(ContextSpec context,
+                            Class<? extends Projection<I, ?, ?>> projectionClass) {
+        return delegate.createProjectionStorage(context, projectionClass);
+    }
+
+    @Override
+    public InboxStorage createInboxStorage(boolean multitenant) {
+        return delegate.createInboxStorage(multitenant);
+    }
+
+    /**
+     * Creates a new {@link EventStore}.
+     *
+     * <p>If the given context does not store events, returns an {@link EmptyEventStore}. Otherwise,
+     * delegates to the wrapped instance.
+     */
+    @Override
+    public EventStore createEventStore(ContextSpec context) {
+        return context.storesEvents()
```

Comment:
Once again, there is a conflict between the name `SystemAware...` and non-system-awareness in the `storesEvents()`. Why not just have `isSystem()` then?